barman/cloud.py

```diff
@@ -1025,26 +1034,58 @@ def __init__(
 
         self.compression = compression
         self.server_name = server_name
-        self.postgres = postgres
         self.cloud_interface = cloud_interface
         self.max_archive_size = max_archive_size
 
         # Stats
         self.copy_start_time = None
         self.copy_end_time = None
 
-    def backup_copy(self, controller, backup_info):
+    def _create_upload_controller(self, backup_id):
+        """
+        Create an upload controller from the specified backup_id
+
+        :param str backup_id: The backup identifier
+        :return: The upload controller
+        :rtype: CloudUploadController
+        """
+        key_prefix = os.path.join(
+            self.cloud_interface.path,
+            self.server_name,
+            "base",
+            backup_id,
+        )
+        return CloudUploadController(
+            self.cloud_interface,
+            key_prefix,
+            self.max_archive_size,
+            self.compression,
+        )
+
+    @abstractmethod
+    def _get_tablespace_location(self, tablespace):
+        """
+        Return the on-disk location of the supplied tablespace
+
+        This will vary depending on whether barman-cloud is running on a live
+        PostgreSQL server or a Barman server
+        """
+
+    def _backup_copy(self, controller, backup_info, pgdata_dir, server_major_version):
         """
         Perform the actual copy of the backup uploading it to cloud storage.
 
         First, it copies one tablespace at a time, then the PGDATA directory,
-        and finally configuration files (if outside PGDATA).
```

Comment:
Yes - the core `_backup_copy` function no longer copies configuration files outside PGDATA because the hook script implementation cannot do that (any such files will already have been copied into the backup dir by `barman backup`). The non-hook-script implementation in CloudBackupUploaderPostgres *does* need to do this, so that logic is moved to [its _backup_copy implementation](https://github.com/EnterpriseDB/barman/pull/353/files/6d6f9a04c6a55e9516e3795fa902805841f1a5ce#diff-3b02e947f51597b51467015d57554983c2b8ad65fdbc77048ccad8d32bcb5cd2R1287-R1319).